File: core-clients/python/ramen_ai/governed_types.py
"""Typed contracts for governed generation requests, responses, and SSE events."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, TypeAlias

GovernedProviderName: TypeAlias = Literal[
    "openai",
    "anthropic",
    "google",
    "synthetic",
    "hyperbolic",
]
GovernedStatusStage: TypeAlias = Literal[
    "accepted",
    "generating",
    "evaluating",
    "regenerating",
]


@dataclass(frozen=True, slots=True)
class GovernedGenerationOptions:
    """Optional provider generation controls."""

    temperature: float | None = None
    max_tokens: int | None = None


@dataclass(frozen=True, slots=True)
class GovernedTokenUsage:
    prompt_tokens: int
    completion_tokens: int
    total_tokens: int


@dataclass(frozen=True, slots=True)
class GovernedAttemptMetadata:
    attempt: int
    provider: str
    model: str
    generation_duration_ms: int
    evaluation_duration_ms: int
    policies_evaluated: int
    allowed: bool
    usage: GovernedTokenUsage | None = None
    rejected_content: str | None = None
    steering_rationale: list[str] | None = None


@dataclass(frozen=True, slots=True)
class GovernedAccounting:
    generation_attempts: int
    evaluation_batches: int
    policy_evaluations: int


@dataclass(frozen=True, slots=True)
class GovernedEvaluationSummary:
    allowed: bool
    policy_ids: tuple[str, ...]
    policies_evaluated: int
    policies_passed: int
    policies_failed: int
    policies_errored: int
    violation_count: int
    statutory_anchors: tuple[str, ...]
    receipt_id: str | None = None


@dataclass(frozen=True, slots=True)
class GovernedCompleteData:
    content: str
    provider: str
    model: str
    attempts: int
    attempt_metadata: tuple[GovernedAttemptMetadata, ...]
    evaluation: GovernedEvaluationSummary
    accounting: GovernedAccounting
    total_duration_ms: int
    usage: GovernedTokenUsage | None = None


@dataclass(frozen=True, slots=True)
class GovernedBlockedData:
    attempts: int
    attempt_metadata: tuple[GovernedAttemptMetadata, ...]
    evaluation: GovernedEvaluationSummary
    accounting: GovernedAccounting
    total_duration_ms: int


@dataclass(frozen=True, slots=True)
class GovernedStatusData:
    stage: GovernedStatusStage
    attempt: int
    violations: int | None = None


@dataclass(frozen=True, slots=True)
class GovernedHeartbeatData:
    timestamp: str


@dataclass(frozen=True, slots=True)
class GovernedStatusEvent:
    event: Literal["status"]
    data: GovernedStatusData


@dataclass(frozen=True, slots=True)
class GovernedHeartbeatEvent:
    event: Literal["heartbeat"]
    data: GovernedHeartbeatData


@dataclass(frozen=True, slots=True)
class GovernedCompleteEnvelope:
    success: Literal[True]
    data: GovernedCompleteData


@dataclass(frozen=True, slots=True)
class GovernedCompleteEvent:
    event: Literal["complete"]
    data: GovernedCompleteEnvelope


GovernedStreamEvent: TypeAlias = (
    GovernedStatusEvent | GovernedHeartbeatEvent | GovernedCompleteEvent
)

File: core-clients/node/src/governed-types.ts
/** Typed contracts for governed generation requests, responses, and SSE events. */

export type GovernedProviderName =
  | "openai"
  | "anthropic"
  | "google"
  | "synthetic"
  | "hyperbolic";

export interface GovernedGenerationOptions {
  temperature?: number;
  maxTokens?: number;
}

export interface GenerateGovernedOptions {
  policyIds?: readonly string[];
  bundleIds?: readonly string[];
  maxRetries?: 0 | 1;
  generation?: GovernedGenerationOptions;
  exposeHealingTrail?: boolean;
}

export interface GovernedGenerateWireRequest {
  prompt: string;
  policy_ids?: string[];
  bundle_ids?: string[];
  max_retries: 0 | 1;
  generation?: {
    temperature?: number;
    max_tokens?: number;
  };
  expose_healing_trail?: boolean;
}

export interface GovernedTokenUsage {
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
}

export interface GovernedAttemptMetadata {
  attempt: number;
  provider: string;
  model: string;
  generation_duration_ms: number;
  evaluation_duration_ms: number;
  policies_evaluated: number;
  allowed: boolean;
  usage?: GovernedTokenUsage;
  rejected_content?: string;
  steering_rationale?: string[];
}

export interface GovernedAccounting {
  generation_attempts: number;
  evaluation_batches: number;
  policy_evaluations: number;
}

export interface GovernedEvaluationSummary {
  allowed: boolean;
  policy_ids: string[];
  policies_evaluated: number;
  policies_passed: number;
  policies_failed: number;
  policies_errored: number;
  violation_count: number;
  statutory_anchors: string[];
  receipt_id?: string;
}

export interface GovernedCompleteData {
  content: string;
  provider: string;
  model: string;
  usage?: GovernedTokenUsage;
  attempts: number;
  attempt_metadata: GovernedAttemptMetadata[];
  evaluation: GovernedEvaluationSummary;
  accounting: GovernedAccounting;
  total_duration_ms: number;
}

export interface GovernedBlockedData {
  attempts: number;
  attempt_metadata: GovernedAttemptMetadata[];
  evaluation: GovernedEvaluationSummary;
  accounting: GovernedAccounting;
  total_duration_ms: number;
}

export type GovernedStatusStage =
  | "accepted"
  | "generating"
  | "evaluating"
  | "scrubbing"
  | "regenerating";

export interface GovernedStatusEvent {
  event: "status";
  data: {
    stage: GovernedStatusStage;
    attempt: number;
    violations?: number;
  };
}

export interface GovernedHeartbeatEvent {
  event: "heartbeat";
  data: {
    timestamp: string;
  };
}

export interface GovernedCompleteEvent {
  event: "complete";
  data: {
    success: true;
    data: GovernedCompleteData;
  };
}

export type GovernedStreamEvent =
  | GovernedStatusEvent
  | GovernedHeartbeatEvent
  | GovernedCompleteEvent;
